src/OrchardCore.Modules/OrchardCore.Media/Shortcodes/ShareShortcodeProvider.cs

```diff
@@ -0,0 +1,174 @@
+    /// prompted to save it using a suggested but not mandatory filename either specified using the "save" arguement, or inferred
+    /// from the content url.  Please note that the short-code argument parser presumes spaces are delimiters, and so content urls
+    /// that include spaces must be quoted; however, this provider will render an encoded url, replacing characters such as spaces
+    /// with their url-hexidecimal format (eg space would be %20).
+    ///
+    /// Additional optional arguements can be specified for styling and accessibility purposes.  These include:
+    /// - tooltip: to create a 'title' attribute on the anchor, and again the shortcode arguement parser presumes a space to be a
+    /// delimiter for the next arguement pair, and so quotes are required for multi-word titles
+    /// - class: to specify the css classes to attach to the element
+    /// </summary>
+    public class ShareShortCodeProvider : IShortcodeProvider
+    {
+
+        private const string FileExtensionSeperator = ".";
+        private static readonly ValueTask<string> Null = new ValueTask<string>((string)null);
+        private static readonly ValueTask<string> ShareShortcode = new ValueTask<string>("[share]");
+        private static readonly HashSet<string> Shortcodes = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "share",
+            "doc",
+            "download"
+        };
+
+        private readonly IMediaFileStore _mediaFileStore;
+        private readonly IHtmlSanitizerService _htmlSanitizerService;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly ResourceManagementOptions _options;
+
+        public ShareShortCodeProvider(
+            IMediaFileStore mediaFileStore,
+            IHtmlSanitizerService htmlSanitizerService,
+            IHttpContextAccessor httpContextAccessor,
+            IOptions<ResourceManagementOptions> options)
+        {
+            _mediaFileStore = mediaFileStore;
+            _htmlSanitizerService = htmlSanitizerService;
+            _httpContextAccessor = httpContextAccessor;
+            _options = options.Value;
+        }
+
+        #region https://stackoverflow.com/a/40361205/549306
+        private static readonly Uri SomeBaseUri = new Uri("http://canbeanything");
+
+        private static string GetFileNameFromUrl(string url)
+        {
+            Uri uri;
+            if (!Uri.TryCreate(url, UriKind.Absolute, out uri))
+            {
+                uri = new Uri(SomeBaseUri, url);
+            }
+
+            return Path.GetFileName(uri.LocalPath);
+        }
+
+        #endregion
+
+        public ValueTask<string> EvaluateAsync(string identifier, Arguments arguments, string content, Context context)
+        {
+            // we can't process what we don't understand, and we are discarding or filtering out anything we don't understand to keep things safe
+            if (!Shortcodes.Contains(identifier))
```

Comment:
A single pre-defined shortcode name is sufficient.
In this case "link".

Reply:
I have made the suggested change and it will be in the resulting submission.

For future reference, there can be value in adopting conventions of other existing systems.  For instance, BBCode assigned the `[url url=?]` or `[url]?[/url]` moniker for a similar purpose, but that shortcode-mechanism doesn't include the download attribute in the generated code that encourages the browser to treat the linked resource differently, so using a different moniker would be entirely appropriate.

[BBCode Reference](https://www.bbcode.org/reference.php)

Reply:
I agree with following existing conventions. BBCode is not very common. `url` in bbcode is ambiguous, and I really think we need both `link` and `url`.